scripts/common/k3s.sh

```diff
@@ -0,0 +1,597 @@
+
+    # TODO(ethan): is this still necessary?
+    # if [ "$CLUSTER_DNS" != "$DEFAULT_CLUSTER_DNS" ]; then
+    #     sed -i "s/$DEFAULT_CLUSTER_DNS/$CLUSTER_DNS/g" /etc/systemd/system/kubelet.service.d/10-kubeadm.conf
+    # fi
+
+    logSuccess "K3S host packages installed"
+}
+
+function k3s_host_packages_ok() {
+    local k3s_version="$1"
+
+    if ! commandExists k3s; then
+        echo "k3s command missing - will install host components"
+        return 1
+    fi
+
+    kubelet --version | grep -q "$(echo $k3s_version | sed "s/-/+/")"
+}
+
+function k3s_get_host_packages_online() {
+    local k3s_version="$1"
+
+    echo "Fetching k3s-${k3s_version}.tar.gz"
+    curl -LO "$DIST_URL/k3s-${k3s_version}.tar.gz"
+    rm -rf $DIR/packages/k3s/${k3s_version} # Cleanup broken/incompatible packages from failed runs
+    tar xf k3s-${k3s_version}.tar.gz
+    rm k3s-${k3s_version}.tar.gz
+}
+
+function k3s_load_images() {
+    local k3s_version="$1"
+
+    logStep "Load K3S images"
+
+    mkdir -p /var/lib/rancher/k3s/agent/images
+    gunzip -c $DIR/packages/k3s/${k3s_version}/assets/k3s-images.linux-amd64.tar.gz > /var/lib/rancher/k3s/agent/images/k3s-images.linux-amd64.tar
+
+    logSuccess "K3S images loaded"
+}
+
+k3s_server_setup_systemd_service() {
+
+    if [ -f "/etc/systemd/system/k3s-server.service" ]; then
+        logSubstep "Systemd service for the K3S Server already exists. Skipping."
+        return
+    fi 
+
+    logStep "Creating K3S Server Systemd Service"
+
+    k3s_create_env_file
+    # Created Systemd unit from https://get.k3s.io/
+    # TODO (dan): check if this should be a server or agent
+    tee /etc/systemd/system/k3s-server.service > /dev/null  <<EOF
+[Unit]
+Description=Lightweight Kubernetes
+Documentation=https://k3s.io
+Wants=network-online.target
+After=network-online.target
+
+[Install]
+WantedBy=multi-user.target
+
+[Service]
+Type=exec
+EnvironmentFile=/etc/systemd/system/k3s-server.env
+KillMode=process
+Delegate=yes
+# Having non-zero Limit*s causes performance problems due to accounting overhead
+# in the kernel. We recommend using cgroups to do container-local accounting.
+LimitNOFILE=1048576
+LimitNPROC=infinity
+LimitCORE=infinity
+TasksMax=infinity
+TimeoutStartSec=0
+Restart=always
+RestartSec=5s
+ExecStartPre=-/sbin/modprobe br_netfilter
+ExecStartPre=-/sbin/modprobe overlay
+ExecStart=/usr/local/bin/k3s \\
+    server
+
+EOF
+    
+    systemctl daemon-reload
+    systemctl start k3s-server.service 
+    systemctl enable k3s-server.service 
+
+    logSuccess "K3S Service created"
+}
+
+k3s_create_env_file() {
+    local fileK3sEnv=/etc/systemd/system/k3s-server.env
+    echo "Creating environment file ${fileK3sEnv}"
+    UMASK=$(umask)
+    umask 0377
+    env | grep '^K3S_' | tee ${fileK3sEnv} >/dev/null
+    env | egrep -i '^(NO|HTTP|HTTPS)_PROXY' | tee -a ${fileK3sEnv} >/dev/null
+    umask $UMASK
+}
+
+k3s_create_symlinks() {
+    local binDir=/usr/local/bin
+
+    for cmd in kubectl crictl ctr; do
```

Comment:
this is cool. would be nice for rke2

Reply:
Seems like the architecture may be different here AFAIK. RKE has actual binaries but[ K3S bundles other tools under their CLI.](https://github.com/k3s-io/k3s/blob/master/pkg/kubectl/main.go) Not sure if it would make as much sense.

I think there is a question if we should install our own tools or not. I don't know the right answer but my first thought it that deferring to rancher's tools will prevent incompat. issues.

Reply:
Got it